Core/src/Geometry/Portal.cpp

```diff
@@ -0,0 +1,306 @@
+    if (bPortal.m_oppositeNormal == nullptr) {
+      ACTS_ERROR(
+          "Portal A has link opposite normal, while b does not. This is not "
+          "supported");
+      throw PortalMergingException();
+    }
+
+    ACTS_VERBOSE("Portals have links opposite normal, merging");
+    mergedOppositeNormal = PortalLinkBase::merge(
+        std::move(aPortal.m_oppositeNormal),
+        std::move(bPortal.m_oppositeNormal), direction, logger);
+  }
+
+  aPortal.m_surface.reset();
+  bPortal.m_surface.reset();
+  return Portal{gctx, std::move(mergedAlongNormal),
+                std::move(mergedOppositeNormal)};
+}
+
+Portal Portal::fuse(const GeometryContext& gctx, Portal& aPortal,
+                    Portal& bPortal, const Logger& logger) {
+  ACTS_DEBUG("Fusing two portals");
+  if (&aPortal == &bPortal) {
+    ACTS_ERROR("Cannot merge a portal with itself");
+    throw PortalMergingException{};
+  }
+
+  bool aHasAlongNormal = aPortal.m_alongNormal != nullptr;
+  bool aHasOppositeNormal = aPortal.m_oppositeNormal != nullptr;
+  bool bHasAlongNormal = bPortal.m_alongNormal != nullptr;
+  bool bHasOppositeNormal = bPortal.m_oppositeNormal != nullptr;
+
+  if (aPortal.m_surface == nullptr || bPortal.m_surface == nullptr) {
+    ACTS_ERROR("Portals have no surface");
+    throw PortalFusingException();
+  }
+
+  if (aPortal.m_surface->associatedDetectorElement() != nullptr ||
+      bPortal.m_surface->associatedDetectorElement() != nullptr) {
+    ACTS_ERROR("Cannot fuse portals with detector elements");
+    throw PortalFusingException();
+  }
+
+  if (!isSameSurface(gctx, *aPortal.m_surface, *bPortal.m_surface)) {
+    ACTS_ERROR("Portals have different surfaces");
+    throw PortalFusingException();
+  }
+
+  if (aPortal.m_surface->surfaceMaterial() != nullptr &&
+      bPortal.m_surface->surfaceMaterial() != nullptr) {
+    ACTS_ERROR("Cannot fuse portals if both have material");
+    throw PortalFusingException();
+  }
+
+  if (aHasAlongNormal == bHasAlongNormal ||
+      aHasOppositeNormal == bHasOppositeNormal) {
+    ACTS_ERROR("Portals have the same links attached");
+    throw PortalFusingException();
+  }
+
+  aPortal.m_surface.reset();
+  bPortal.m_surface.reset();
+  if (aHasAlongNormal) {
+    ACTS_VERBOSE("Taking along normal from lhs, opposite normal from rhs");
+    return Portal{gctx, std::move(aPortal.m_alongNormal),
+                  std::move(bPortal.m_oppositeNormal)};
+  } else {
+    ACTS_VERBOSE("Taking along normal from rhs, opposite normal from lhs");
+    return Portal{gctx, std::move(bPortal.m_alongNormal),
+                  std::move(aPortal.m_oppositeNormal)};
+  }
+}
+
+bool Portal::isSameSurface(const GeometryContext& gctx, const Surface& a,
+                           const Surface& b) {
+  if (&a == &b) {
+    return true;
+  }
+
+  if (a.type() != b.type()) {
+    return false;
+  }
+
+  if (a.bounds() != b.bounds()) {
+    return false;
+  }
+
+  if (!a.transform(gctx).isApprox(b.transform(gctx), 1e-9)) {
```

Comment:
I wonder if we should rather use `==` or a common global epsilon. Do you have a case in mind where the transform might be numerically altered compared to the other one?

Reply:
I don't think we should use `==`, because if volumes are being connected that come from size harmonization, I'm not willing to guarantee that they have **exactly** the same bounds and hence boundary surfaces.

I can make the `10e-9` a named variable, if you think that's more clear.

Reply:
I think it is ok as it is. I just wondered if we want to come up with a global constant for this at some point in case we hard code different magic numbers for the same purpose in different places. Similar to `s_onSurfaceTolerance`

Reply:
I added `s_transformEquivalentTolerance` to `Algebra.hpp` in this PR for this.